### PR TITLE
Put pod full name, namespace and uid into docker label

### PR DESCRIPTION
**Pull request by Random-Liu**
This PR mainly did 3 things:
1. Add file container_labels.go, and move label and filter functions into it, so as to simplify  the label related code and prepare for adding more labels in the future.
2. Add pod full name filter in GetPodStatus() so that ListContainers() will only list containers belong to the target pod.
3. Add related test case and make fakeDockerContainer support label filter.

Some questions or TODOs:
1. A lot of manually initialized containers in old test cases of dockertools don't have docker labels, if they use functions using label filter (current now only GetPodStatus() uses label filter), the test case may get unexpected result. I added a simple function to set default labels, but it is still inconvenient to add label for each test case. In this PR I just modified test cases which went wrong. Should I add label for each test case? Or should I think a better way?
2. In fact, pod full name label was added before https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/dockertools/manager.go#L701. Can I assume that even old containers have pod name label? If not, I should comment my filter code and uncomment it in the future.

(For issue #15089)

### CODE REVIEW

**Comment by ncdc** [comment]
@kubernetes/rh-cluster-infra @mfojtik @smarterclayton

**Comment by Random-Liu** [comment]
@k8s-bot test this please.

**Comment by Random-Liu** [comment]
@k8s-bot test this please.

**Comment by ashcrow** [comment]
Shippable failure is likely a flake:
```
FAIL	k8s.io/kubernetes/pkg/storage/etcd	14.651s
```

**Comment by Random-Liu** [comment]
@k8s-bot test this please.

**Comment by Random-Liu** [comment]
Yeah, maybe. I passed the test locally.

**Comment by Random-Liu** [comment]
@ashcrow Thanks! :)

**Comment by Random-Liu** [comment]
@k8s-bot test this please.

File: pkg/kubelet/dockertools/docker.go [hunk 1]
```diff
@@ -211,6 +211,7 @@ func (p throttledDockerPuller) IsImagePresent(name string) (bool, error) {
 	return p.puller.IsImagePresent(name)
 }
 
+// TODO (random-liu) Almost never used, should we remove this?
```
Comment (yujuhong) [inline pkg/kubelet/dockertools/docker.go]: This is used by `GetKubeletDockerContainers`. Do you want to rewrite the function?

File: pkg/kubelet/dockertools/container_labels.go [hunk 2]
```diff
@@ -0,0 +1,79 @@
+/*
```
Comment (yujuhong) [inline pkg/kubelet/dockertools/container_labels.go]: I think the"labels.go" is clear enough that we don't need to add the "container_" prefix :)

File: pkg/kubelet/dockertools/container_labels.go [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+import (
+	"strconv"
+
+	docker "github.com/fsouza/go-dockerclient"
+	"github.com/golang/glog"
+	"k8s.io/kubernetes/pkg/types"
+)
+
+const (
+	kubernetesPodNameLabel                = "io.kubernetes.pod.name"
```
Comment (yujuhong) [inline pkg/kubelet/dockertools/container_labels.go]: We agreed to split this to Name and Namespace. Are you going to do it in a separate PR?

Comment (yujuhong) [inline pkg/kubelet/dockertools/container_labels.go]: I see you already fixed this. Thanks!

**Comment by yujuhong** [comment]
@Random-Liu, could you remove the "label filtering" code from this PR? I think this PR should focus on writing the labels :-)

File: pkg/kubelet/dockertools/container_labels.go [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+	docker "github.com/fsouza/go-dockerclient"
+	"github.com/golang/glog"
+	"k8s.io/kubernetes/pkg/types"
+)
+
+const (
+	kubernetesPodNameLabel      = "io.kubernetes.pod.name"
+	kubernetesPodNamespaceLabel = "io.kubernetes.pod.namespace"
+
+	kubernetesPodLabel                    = "io.kubernetes.pod.data"
```
Comment (vishh) [inline pkg/kubelet/dockertools/container_labels.go]: What is this meant for? It is not clear from the variable name. Can you add comments?

Comment (yujuhong) [inline pkg/kubelet/dockertools/container_labels.go]: @vishh, @Random-Liu didn't add this. This was added for PreStop hook and the label contains the entire serialized pod... I want to remove this. See #/14768 for the details.

File: pkg/kubelet/dockertools/docker.go (hunk 1 above)
Comment (Random-Liu) [inline pkg/kubelet/dockertools/docker.go]: Ha, what I mean is that we defined the the type "DockerContainers", and defined the function "FindPodContainer". But it seems that we just use "DockerContainers" as slice of APIcontainers (https://github.com/kubernetes/kubernetes/search?utf8=%E2%9C%93&q=GetKubeletDockerContainers&type=Code), never use function like FindPodContainer. So I just wonder whether we should remove this type, :)

Comment (Random-Liu) [inline pkg/kubelet/dockertools/docker.go]: Just curious when reading the code, so add a comment here. :)

File: pkg/kubelet/dockertools/container_labels.go (hunk 2 above)
Comment (Random-Liu) [inline pkg/kubelet/dockertools/container_labels.go]: OK! I'll do it right now.

**Comment by Random-Liu** [comment]
OK. I'll remove the filter related code in this PR. :)

File: pkg/kubelet/dockertools/docker.go (hunk 1 above)
Comment (yujuhong) [inline pkg/kubelet/dockertools/docker.go]: I believe it was used in other places but we gradually migrated off. Personally I'd be happy to see the type be removed. Feel free to send a PR for this :)

File: pkg/kubelet/dockertools/container_labels.go (hunk 4 above)
Comment (Random-Liu) [inline pkg/kubelet/dockertools/container_labels.go]: I read code about this yesterday, it seems that one of the options is to add some "better-formatted" labels on each container (rather than add the whole pod data), for example add a label like "container_metadata" and write all needed data like PreStop handler into it...

File: pkg/kubelet/dockertools/docker.go (hunk 1 above)
Comment (Random-Liu) [inline pkg/kubelet/dockertools/docker.go]: OK. :) Thanks! Haha~

File: pkg/kubelet/dockertools/labels.go [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+			glog.Errorf("Parsing restart count string %s error: %v", restartCountString, err)
+			// Just set restartCount to 0 to handle this abnormal case
+			restartCount = 0
+		}
+		return restartCount
+	} else {
+		// Get restartCount from docker label. If there is no restart count label in an container,
+		// it should be an old container or an invalid container, we just set restart count to 0.
+		// Don't report error, because there must be many old containers without the label now
+		glog.V(3).Infof("Container doesn't have label %s, it may be an old or invalid container", kubernetesContainerRestartCountLabel)
```
Comment (yujuhong) [inline pkg/kubelet/dockertools/labels.go]: For the logging message to be meaningful in anyway, we need the container id. I'd suggest returning (int, error) and let the caller handle the error.

File: pkg/kubelet/dockertools/manager.go [hunk 6]
```diff
@@ -1643,8 +1620,8 @@ func (dm *DockerManager) createPodInfraContainer(pod *api.Pod) (kubetypes.Docker
 		return "", err
 	}
 
-	// There is no meaningful labels for infraContainer now, so just pass nil.
-	id, err := dm.runContainerInPod(pod, container, netNamespace, getIPCMode(pod), getPidMode(pod), nil)
+	// Current now we don't care about restart count of infra container, just set it to 0.
```
Comment (yujuhong) [inline pkg/kubelet/dockertools/manager.go]: Remove "Current now"

FYI, you can use "currently" or not use it at all (implied). "Current now" is not grammatically correct :-)

**Comment by yujuhong** [comment]
lgtm with some nits :)

File: pkg/kubelet/dockertools/manager.go (hunk 6 above)
Comment (Random-Liu) [inline pkg/kubelet/dockertools/manager.go]: Ops...I don't even know where I learned it. I really need more suggestions like this about my grammar, words etc. Thank you very much! :)

File: pkg/kubelet/dockertools/labels.go (hunk 5 above)
Comment (Random-Liu) [inline pkg/kubelet/dockertools/labels.go]: Ok!

**Comment by yujuhong** [comment]
lgtm. please squash.

**Comment by Random-Liu** [comment]
Sure. Thanks :)

**Comment by Random-Liu** [comment]
Squashed. :)

**Comment by Random-Liu** [comment]
@k8s-bot test this

**Comment by a-robinson** [comment]
@k8s-bot unit test this
